**Context**

There is a huge demand for used cars in the Indian Market today. As sales of new cars have slowed down in the recent past, the pre-owned car market has continued to grow over the past years and is larger than the new car market now. Cars4U is a budding tech start-up that aims to find footholes in this market.

In 2018-19, while new car sales were recorded at 3.6 million units, around 4 million second-hand cars were bought and sold. There is a slowdown in new car sales and that could mean that the demand is shifting towards the pre-owned market. In fact, some car sellers replace their old cars with pre-owned cars instead of buying new ones. Unlike new cars, where price and supply are fairly deterministic and managed by OEMs (Original Equipment Manufacturer / except for dealership level discounts which come into play only in the last stage of the customer journey), used cars are very different beasts with huge uncertainty in both pricing and supply. Keeping this in mind, the pricing scheme of these used cars becomes important in order to grow in the market.

As a senior data scientist at Cars4U, you have to come up with a pricing model that can effectively predict the price of used cars and can help the business in devising profitable strategies using differential pricing. For example, if the business knows the market price, it will never sell anything below it. 

**Data**

- S.No. : Serial Number
- Name : Name of the car which includes Brand name and Model name
- Location : The location in which the car is being sold or is available for purchase Cities
- Year : Manufacturing year of the car
- Kilometers_driven : The total kilometers driven in the car by the previous owner(s) in KM.
- Fuel_Type : The type of fuel used by the car. (Petrol, Diesel, Electric, CNG, LPG)
- Transmission : The type of transmission used by the car. (Automatic / Manual)
- Owner : Type of ownership
- Mileage : The standard mileage offered by the car company in kmpl or km/kg
- Engine : The displacement volume of the engine in CC.
- Power : The maximum power of the engine in bhp.
- Seats : The number of seats in the car.
- New_Price : The price of a new car of the same model in INR Lakhs.(1 Lakh = 100, 000)
- Price : The price of the used car in INR Lakhs (1 Lakh = 100, 000)

**Objective**

- Explore and visualize the dataset.
- Build a linear regression model to predict the prices of used cars.
- Generate a set of insights and recommendations that will help the business.

## Data structure and initial observations

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import *
from sklearn.linear_model import *

%matplotlib inline

In [5]:
data = pd.read_csv('used_cars_data.csv')   # the original data

In [6]:
df = data.copy()   # copy the data to df

In [7]:
df.head()   # display the first 5 rows

S.No.                              Name    Location  Year  \
0      0            Maruti Wagon R LXI CNG      Mumbai  2010   
1      1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
2      2                      Honda Jazz V     Chennai  2011   
3      3                 Maruti Ertiga VDI     Chennai  2012   
4      4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              72000       CNG       Manual      First  26.6 km/kg   998 CC   
1              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
2              46000    Petrol       Manual      First   18.2 kmpl  1199 CC   
3              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
4              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   

       Power  Seats  New_Price  Price  
0  58.16 bhp    5.0        NaN   1.75  
1  126.2 bhp    5.0        NaN  12.50  
2   88.7 bhp    5.0  8.61 Lakh   4.50  
3  88.76 bhp    7.0        NaN   6.00  
4  140.8 bhp    5.0        NaN  17.74

In [12]:
df.sample(10)   # randomly display 10 rows

S.No.                                               Name   Location  \
1421   1421                                 Maruti Celerio VXI      Delhi   
3362   3362                                   BMW X1 sDrive20d     Mumbai   
3873   3873                                 Maruti Celerio ZXI    Chennai   
7139   7139                      Hyundai i20 1.2 Sportz Option    Kolkata   
5264   5264                                    Audi A4 2.0 TDI      Kochi   
5599   5599  Mercedes-Benz E-Class 2009-2013 E 220 CDI Avan...     Jaipur   
3464   3464                                    Maruti Ritz VDi  Hyderabad   
4879   4879                             Maruti Swift Dzire ZXI     Mumbai   
3077   3077                      Tata Tiago Wizz 1.05 Revotorq  Hyderabad   
984     984                      BMW 5 Series 520d Luxury Line     Mumbai   

      Year  Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
1421  2014              77000    Petrol       Manual      First   23.1 kmpl   
3362  2013              43000    Diesel    Automatic      First  17.05 kmpl   
3873  2015              57818    Petrol       Manual      First   23.1 kmpl   
7139  2012              36200    Petrol       Manual      First   17.0 kmpl   
5264  2014              55730    Diesel    Automatic      First  16.55 kmpl   
5599  2013              60000    Diesel    Automatic      First   10.0 kmpl   
3464  2010             110471    Diesel       Manual      First   21.1 kmpl   
4879  2014              48241    Petrol       Manual      First   19.1 kmpl   
3077  2017              53000    Diesel       Manual      First  27.28 kmpl   
984   2012              61240    Diesel    Automatic      First  22.48 kmpl   

       Engine       Power  Seats   New_Price  Price  
1421   998 CC   67.04 bhp    5.0         NaN   2.80  
3362  1995 CC     184 bhp    5.0         NaN  17.50  
3873   998 CC   67.04 bhp    5.0         NaN   4.30  
7139  1197 CC      80 bhp    5.0         NaN    NaN  
5264  1968 CC  147.51 bhp    5.0         NaN  15.21  
5599  2148 CC     170 bhp    5.0         NaN  27.00  
3464  1248 CC    73.9 bhp    5.0         NaN   3.50  
4879  1197 CC    85.8 bhp    5.0         NaN   4.50  
3077  1047 CC      69 bhp    5.0         NaN   5.00  
984   1995 CC     190 bhp    5.0  70.43 Lakh  17.50

In [8]:
df.shape 

(7253, 14)

In [9]:
df.info()    # data types

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7253 entries, 0 to 7252
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   S.No.              7253 non-null   int64  
 1   Name               7253 non-null   object 
 2   Location           7253 non-null   object 
 3   Year               7253 non-null   int64  
 4   Kilometers_Driven  7253 non-null   int64  
 5   Fuel_Type          7253 non-null   object 
 6   Transmission       7253 non-null   object 
 7   Owner_Type         7253 non-null   object 
 8   Mileage            7251 non-null   object 
 9   Engine             7207 non-null   object 
 10  Power              7207 non-null   object 
 11  Seats              7200 non-null   float64
 12  New_Price          1006 non-null   object 
 13  Price              6019 non-null   float64
dtypes: float64(2), int64(3), object(9)
memory usage: 793.4+ KB


**observation:** 
- 7253 rows x 14 columns. Serial number, year, Kilometers_Driven, Seats and price are numerical. Others are objects.
- Name of the car includes Brand name and Model name, which seem to be seperated into two columns.
- Mileage, engine and Power should be numbers instead of objects.
- Mixed measurements for mileage.
- Lots of missing values in New_Price column.
- serial number is identical with the index. So it could be used for index instead.

In [10]:
df.isnull().sum().sort_values(ascending = False)    # check missing values

New_Price            6247
Price                1234
Seats                  53
Engine                 46
Power                  46
Mileage                 2
S.No.                   0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
dtype: int64

**observation:**
- New Price, Price, Seats, Engine, Power and Mileage column have missing values.
- The number of missing values in New_Price is 6247, taking up more than 85% of the total 7253 observations. It is impossible to fix by imputation. It makes more sense to just drop off this column.

## Data preprocessing

In [16]:
df.set_index('S.No.', inplace = True)  # set serial number as index column
df

Name    Location  Year  \
S.No.                                                                        
0                                 Maruti Wagon R LXI CNG      Mumbai  2010   
1                       Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
2                                           Honda Jazz V     Chennai  2011   
3                                      Maruti Ertiga VDI     Chennai  2012   
4                        Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
...                                                  ...         ...   ...   
7248                   Volkswagen Vento Diesel Trendline   Hyderabad  2011   
7249                              Volkswagen Polo GT TSI      Mumbai  2015   
7250                              Nissan Micra Diesel XV     Kolkata  2012   
7251                              Volkswagen Polo GT TSI        Pune  2013   
7252   Mercedes-Benz E-Class 2009-2013 E 220 CDI Avan...       Kochi  2014   

       Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
S.No.                                                                    
0                  72000       CNG       Manual      First  26.6 km/kg   
1                  41000    Diesel       Manual      First  19.67 kmpl   
2                  46000    Petrol       Manual      First   18.2 kmpl   
3                  87000    Diesel       Manual      First  20.77 kmpl   
4                  40670    Diesel    Automatic     Second   15.2 kmpl   
...                  ...       ...          ...        ...         ...   
7248               89411    Diesel       Manual      First  20.54 kmpl   
7249               59000    Petrol    Automatic      First  17.21 kmpl   
7250               28000    Diesel       Manual      First  23.08 kmpl   
7251               52262    Petrol    Automatic      Third   17.2 kmpl   
7252               72443    Diesel    Automatic      First   10.0 kmpl   

        Engine      Power  Seats  New_Price  Price  
S.No.                                               
0       998 CC  58.16 bhp    5.0        NaN   1.75  
1      1582 CC  126.2 bhp    5.0        NaN  12.50  
2      1199 CC   88.7 bhp    5.0  8.61 Lakh   4.50  
3      1248 CC  88.76 bhp    7.0        NaN   6.00  
4      1968 CC  140.8 bhp    5.0        NaN  17.74  
...        ...        ...    ...        ...    ...  
7248   1598 CC  103.6 bhp    5.0        NaN    NaN  
7249   1197 CC  103.6 bhp    5.0        NaN    NaN  
7250   1461 CC   63.1 bhp    5.0        NaN    NaN  
7251   1197 CC  103.6 bhp    5.0        NaN    NaN  
7252   2148 CC    170 bhp    5.0        NaN    NaN  

[7253 rows x 13 columns]